File: app/src/main/java/org/citra/citra_emu/applets/MiiSelector.java
package org.citra.citra_emu.applets;

/**
 * Mii 选择 applet —— 与上游 org.citra.citra_emu.applets.MiiSelector 一致：
 * native 构造 MiiSelectorConfig（按字段名 enableCancelButton/
 * initiallySelectedMiiIndex/miiNames/title 写入），调用 static Execute(config)
 * 阻塞等待，返回 MiiSelectorData（returnCode: 0=选定, 1=取消）。
 *
 * NesStation 简化实现：直接返回「取消」—— 绝大多数游戏在取消时回退
 * 系统 Mii（game 侧不影响运行）。返回数据字段名 returnCode/index 被
 * native 直接读取（GetLongField/GetIntField），勿改名。
 */
public final class MiiSelector {

    public static final MiiSelector INSTANCE = new MiiSelector();

    private MiiSelector() {
    }

    public static final class MiiSelectorConfig
        implements java.io.Serializable {
        private boolean enableCancelButton;
        private long initiallySelectedMiiIndex;
        public String[] miiNames;
        private String title;

        public MiiSelectorConfig() {
        }

        public final boolean getEnableCancelButton() {
            return enableCancelButton;
        }

        public final void setEnableCancelButton(boolean v) {
            this.enableCancelButton = v;
        }

        public final long getInitiallySelectedMiiIndex() {
            return initiallySelectedMiiIndex;
        }

        public final void setInitiallySelectedMiiIndex(long v) {
            this.initiallySelectedMiiIndex = v;
        }

        public final String[] getMiiNames() {
            return miiNames;
        }

        public final void setMiiNames(String[] v) {
            this.miiNames = v;
        }

        public final String getTitle() {
            return title;
        }

        public final void setTitle(String v) {
            this.title = v;
        }
    }

    public static final class MiiSelectorData {
        private long returnCode;
        private int index;

        public MiiSelectorData(long returnCode, int index) {
            this.returnCode = returnCode;
            this.index = index;
        }

        public final long getReturnCode() {
            return returnCode;
        }

        public final void setReturnCode(long v) {
            this.returnCode = v;
        }

        public final int getIndex() {
            return index;
        }

        public final void setIndex(int v) {
            this.index = v;
        }
    }

    /** native 调用：返回取消（returnCode=1），游戏回退默认 Mii。 */
    public static MiiSelectorData Execute(MiiSelectorConfig config) {
        android.util.Log.i("CitraHost", "MiiSelector.Execute -> cancel (returnCode=1)");
        return new MiiSelectorData(1, 0);
    }
}

File: app/src/main/java/org/citra/citra_emu/applets/MiiSelector.kt
package org.citra.citra_emu.applets

import android.util.Log
import androidx.annotation.Keep

/**
 * Azahar Mii 选择器小程序 —— JNI 契约桩。
 *
 * 原生侧（jni/applets/mii_selector.cpp）在 JNI_OnLoad（IDCache 初始化）时缓存
 * `Execute(Lorg/citra/citra_emu/applets/MiiSelector$MiiSelectorConfig;)Lorg/citra/citra_emu/applets/MiiSelector$MiiSelectorData;`
 * —— 注意是 **PascalCase 的 `Execute`**（AzaharPlus fork 的实际契约，经 libazahar.so
 * 反汇编验证；上游官方 azahar 亦为 `Execute`），并按字段名反射读写 MiiSelectorConfig
 * （enableCancelButton:Z / title:Ljava/lang/String; / initiallySelectedMiiIndex:J /
 * miiNames:[Ljava/lang/String;）与 MiiSelectorData（returnCode:J / index:I）。
 * 最小实现：取消（returnCode = 1，与上游 CANCEL 常量一致）。
 */
@Keep
object MiiSelector {

    const val RETURN_CODE_CANCEL: Long = 1L
    const val RETURN_CODE_OK: Long = 0L

    class MiiSelectorConfig {
        @JvmField
        var enableCancelButton: Boolean = false

        @JvmField
        var title: String? = null

        @JvmField
        var initiallySelectedMiiIndex: Long = 0L

        @JvmField
        var miiNames: Array<String?> = arrayOfNulls(0)
    }

    class MiiSelectorData {
        @JvmField
        var returnCode: Long = RETURN_CODE_CANCEL

        @JvmField
        var index: Int = 0
    }

    /**
     * 方法名必须为 `Execute`（PascalCase）—— JNI_OnLoad 按 GetStaticMethodID
     * 字符串名查找，camelCase 的 `execute` 会抛 NoSuchMethodError 并因 pending
     * exception 直接 SIGABRT 闪退（应用启动即崩，无法恢复）。
     */
    @JvmStatic
    fun Execute(config: MiiSelectorConfig?): MiiSelectorData {
        Log.i("AzaharNative", "MiiSelector executed (cancel stub)")
        return MiiSelectorData().apply { returnCode = RETURN_CODE_CANCEL }
    }
}
